shepherd_bms/src/main.cpp

```diff
@@ -243,78 +242,113 @@ uint32_t faultCheck(AccumulatorData_t *accData)
+        {
+            overChgCurrTimer.cancelTimer();
+            faultEvalState = BEFORE_TIMER_START;
+        }
+    }
 
 	// Low cell voltage fault
-	if (accData->minVoltage.val < MIN_VOLT * 10000) 
-	{
-
-		underVoltCount++;
-		if (underVoltCount > 900)
-		{ // 9 seconds @ 100Hz rate
-			faultStatus |= CELL_VOLTAGE_TOO_LOW;
-		}
-	} 
-	else 
-	{
-		underVoltCount = 0;
-	}
+    if (faultEvalState == BEFORE_TIMER_START && accData->minVoltage.val < MIN_VOLT * 10000)
+    {
+        underVoltTimer.startTimer(UNDER_VOLT_TIME );
+        faultEvalState = FAULT_EVAL;
+    }
+    else if (faultEvalState == FAULT_EVAL)
+    {
+        if (underVoltTimer.isTimerExpired())
+        {
+            faultStatus |= CELL_VOLTAGE_TOO_LOW;
+        }
+        else if (!(accData->minVoltage.val < MIN_VOLT * 10000))
+        {
+            underVoltTimer.cancelTimer();
+            faultEvalState = BEFORE_TIMER_START;
+        }
+    }
 
 	// High cell voltage fault
-	if (((accData->maxVoltage.val > MAX_VOLT * 10000) && digitalRead(CHARGE_DETECT) == HIGH) || (accData->maxVoltage.val > MAX_CHARGE_VOLT * 10000)) 
-	{ // Needs to be reimplemented with a flag for every cell in case multiple go over
-		overVoltCount++;
-		if (overVoltCount > 900) { // 9 seconds @ 100Hz rate
-			faultStatus |= CELL_VOLTAGE_TOO_HIGH;
-		}
-	} 
-	else 
-	{
-		overVoltCount = 0;
-	}
+    if (faultEvalState == BEFORE_TIMER_START && ((accData->maxVoltage.val > MAX_VOLT * 10000) && digitalRead(CHARGE_DETECT) == HIGH) || (accData->maxVoltage.val > MAX_CHARGE_VOLT * 10000))
+    {
+        overVoltTimer.startTimer(OVER_VOLT_TIME);
+        faultEvalState = FAULT_EVAL;
+    }
+    else if (faultEvalState == FAULT_EVAL)
+    {
+        if (overVoltTimer.isTimerExpired())
+        {
+            faultStatus |= CELL_VOLTAGE_TOO_HIGH;
+        }
+        else if (!((accData->maxVoltage.val > MAX_VOLT * 10000) && digitalRead(CHARGE_DETECT) == HIGH) || (accData->maxVoltage.val > MAX_CHARGE_VOLT * 10000))
+        {
+            overVoltTimer.cancelTimer();
+            faultEvalState = BEFORE_TIMER_START;
+        }
+    }
 
 	// High Temp Fault
 	if (accData->maxTemp.val > MAX_CELL_TEMP) {
 		faultStatus |= PACK_TOO_HOT;
 	}
 
 	// Extremely low cell voltage fault
-	if (accData->minVoltage.val < 900) 
-	{ // 90mV
-		lowCellCount++;
-		if (lowCellCount > 100) { // 1 seconds @ 100Hz rate
-			faultStatus |= LOW_CELL_VOLTAGE;
-		}
-	} 
-	else 
-	{
-		lowCellCount = 0;
-	}
+    if (faultEvalState == BEFORE_TIMER_START && (accData->minVoltage.val < 900))
+    {
+        lowCellTimer.startTimer(LOW_CELL_TIME);
+        faultEvalState = FAULT_EVAL;
+    }
+    else if (faultEvalState == FAULT_EVAL)
+    {
+        if (lowCellTimer.isTimerExpired())
+        {
+            faultStatus |= LOW_CELL_VOLTAGE;
+        }
+        else if (!(accData->minVoltage.val < 900))
+        {
+            lowCellTimer.cancelTimer();
+            faultEvalState = BEFORE_TIMER_START;
+        }
+    }
```

Comment:
so just a general concern: 

we can be faulting in multiple ways at once (this is why the fault status updates with the bitwise "or" operator instead of just being set equal ) Right now, the fault eval enum is the same for all faults. So, if one fault hits, itll cause the faultEvalState to go to "FAULT_EVAL"  and be recorded this way for all faults. This may mean we need to straight up make enum objects for all the different faults so they can get their own status, but complexity starts to get really high with that. Will consider other options in the meantime to see if that can be avoided

Reply:
Is this still the case? It seems like there is a separate one now for every type of fault

Reply:
Nvm Im dumb. I see there was a commit to fix it